cloudevents/formats/avro-turbo-format.md

```diff
@@ -0,0 +1,217 @@
+| Integer       | [int][avro-primitives]                                                 |
+| String        | [string][avro-primitives]                                              |
+| Binary        | [bytes][avro-primitives]                                               |
+| URI           | [string][avro-primitives] following [RFC 3986 §4.3][rfc3986-section43] |
+| URI-reference | [string][avro-primitives] following [RFC 3986 §4.1][rfc3986-section41] |
+| Timestamp     | [long][avro-primitives]                                                |
+
+Extension specifications MAY define secondary mapping rules for the values of
+attributes they define, but MUST also include the previously defined primary
+mapping.
+
+### 2.3 OPTIONAL Attributes
+
+CloudEvents Spec defines OPTIONAL attributes. The Avro format defines that these
+fields MUST use the `null` type and the actual type through the
+[union][avro-unions].
+
+Example:
+
+```json
+["null", "string"]
+```
+
+### 2.4 Definition
+
+Users of Avro MUST use a message whose binary encoding is identical to the one
+described by the [CloudEvent Avro Turbo Schema](cloudevents-turbo.avsc):
+
+```json
+{
+  "namespace": "io.cloudevents.v1.avro",
+  "type": "record",
+  "name": "CloudEvent",
+  "version": "1.0",
+  "doc": "Avro Turbo Event Format for CloudEvents",
+  "fields": [
+    {
+      "name": "id",
+      "type": "string"
+    },
+    {
+      "name": "source",
+      "type": "string"
+    },
+    {
+      "name": "type",
+      "type": "string"
+    },
+    {
+      "name": "datacontenttype",
+      "type": [
+        "null",
+        "string"
+      ],
+      "default": null
+    },
+    {
+      "name": "dataschema",
+      "type": [
+        "null",
+        "string"
+      ],
+      "default": null
+    },
+    {
+      "name": "subject",
+      "type": [
+        "null",
+        "string"
+      ],
+      "default": null
+    },
+    {
+      "name": "time",
+      "type": [
+        "null",
+        {
+          "type": "long",
+          "logicalType": "timestamp-millis"
+        }
+      ],
+      "default": null
+    },
+    {
+      "name": "attribute",
+      "type": {
+        "type": "map",
+        "values": [
+          "null",
+          "boolean",
+          "int",
+          {
+            "type": "long",
+            "logicalType": "timestamp-millis"
+          },
+          "string",
+          "bytes"
+        ]
+      },
+      "default": {}
+    },
+    {
+      "name": "data",
+      "type": [
+        "bytes",
+        "null"
+      ],
+      "default": "null"
+    }
+  ]
+}
+```
+
+## 3 Data
+
+Before encoding, the AVRO serializer MUST first determine the runtime data type
+of the content. This can be determined by examining the data for invalid UTF-8
+sequences or by consulting the `datacontenttype` attribute.
+
+## 4 Transport
+
+Transports that support content identification MUST use the following designation:
+
+```text
+ application/cloudevents+avro-turbo
```

Comment:
I"m assuming we need to do this because avro doesn't (as far as I can tell) encode the namespace into the serialized representation ?

There may be some pushback on using a non-registered suffix, but given I don't think application/cloudevents is registered either it may not be an issue ...

Reply:
I can't find details in RFC 2046 about what the use of "+" is intended to mean in a subtype...

Reply:
I'm assuming this is not an issue, because it is widely used in CE.